api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java

```diff
@@ -83,28 +88,29 @@ class OAuth2Client implements OAuth2Authenticator, Closeable {
     defaultAccessTokenLifespan = params.getDefaultAccessTokenLifespan();
     defaultRefreshTokenLifespan = params.getDefaultRefreshTokenLifespan();
     refreshSafetyWindow = params.getRefreshSafetyWindow();
+    idleInterval = params.getIdleInterval();
     tokenExchangeEnabled = params.getTokenExchangeEnabled();
     httpClient = params.getHttpClient().addResponseFilter(this::checkErrorResponse).build();
-    executor = params.getExecutor().orElseGet(OAuth2Client::createDefaultExecutor);
-    shouldCloseExecutor = !params.getExecutor().isPresent();
+    executor = params.getExecutor();
+    shouldCloseExecutor = executor instanceof OAuth2TokenRefreshExecutor;
     objectMapper = params.getObjectMapper();
+    clock = params.getClock();
+    lastAccess = clock.get();
     currentTokensStage =
-        started.thenApplyAsync((v) -> fetchNewTokens(), executor).whenComplete(this::log);
-    currentTokensStage.thenAccept(this::scheduleTokensRenewal);
-  }
-
-  private static ScheduledExecutorService createDefaultExecutor() {
-    return Executors.newSingleThreadScheduledExecutor(
-        runnable -> {
-          Thread thread = Executors.defaultThreadFactory().newThread(runnable);
-          thread.setDaemon(true);
-          thread.setName("oauth2-client");
-          return thread;
-        });
+        started
+            .thenApplyAsync((v) -> fetchNewTokens(), executor)
+            .whenComplete((tokens, error) -> this.log(error));
+    currentTokensStage.thenAccept(this::maybeScheduleTokensRenewal);
   }
 
   @Override
   public AccessToken authenticate() {
+    Instant now = clock.get();
+    lastAccess = now;
+    if (sleeping.compareAndSet(true, false)) {
+      LOGGER.debug("Waking up...");
+      scheduleOrExecuteTokensRenewal(getCurrentTokens(), now, Duration.ZERO);
```

Comment:
if this throws, should we set `sleeping` back to `true` to force a renewal on the next `authenticate()` call?

Reply:
The only way I can image where this would throw is if `executor.schedule` throws `RejectedExecutionException`. I prefer for this exception to propagate in fact, as the OAuth2 client is probably left in an usable state.

But you are remark is excellent; and I agree that we might want to revisit this and strengthen things a bit: e.g. if `executor.schedule` throws `RejectedExecutionException` inside the bakground thread, the error will be swallowed here:

https://github.com/projectnessie/nessie/blob/98448b542fd8c9f2f69a6dbeb265c3d94c6bf21c/api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java#L212

That's indeed bad and the client would be left in an unrecoverable state for sure. I think we should fail the `currentTokensStage` in that case, so that the exception gets propagated to client code when the user calls `authenticate()`.

Do you mind if I do this in another PR? As it's only partially related to sleeping/waking up.